### ISSUE TITLE
ReflectionOnlyLoad and ReflectionOnlyLoadFrom

### ISSUE THREAD

**Issue by oznetmaster** [nunit/nunit#654]
These Assembly class methods are called in nunit.engine, nunit.engine.api, and mono-addins.

They load the specified assembly into the reflection-only context of the current AppDomain.  Assemblies that are loaded in this way remain in memory until the AppDomain is unloaded, which in all of these cases appears to be when the engine terminates.

This seemingly can result in the same assembly being loaded twice (or more?) into memory.  If it is only probed, and never actually loaded into an execution context for some reason, it just sits there using memory.

These RelectionOnlyLoad methods are called to obtain very small amounts of information from the assembly.  Version information, assembly references.

In the CF version of the core-engine, I had to replace these calls because CF does not support them.  I initially used a stripped down version of mono.cecil, but have recently found a different open source library (asmex.fileviewer) that does the same thing and is even smaller.

Using this library (or any similar one) is required for the CF build.  However, I am wondering if this should be used for all builds.  The reason is that it is file based.  It opens the assembly as a file, reads the necessary information, and then closes the file.  The information itself is totally disposable after it is used.  Hence, no potential memory leaks or memory bloat.

I am further stripping the library that I am using to only provide the exact information required by NUnit, so it should get even smaller and faster.  Currently, it is only about 30K + data.

(Actually nunit.engine contains a class AssemblyReader which does this for some header information, but not all necessary information).

**Comment by CharliePoole** [comment]
Yes, I have been thinking about that as well. It's a good point.

I have to confess that when I first used ReflectionOnlyLoad, I was under the impression that assemblies loaded this way could actually be unloaded. Not so, as you point out! The domain used has to be unloaded and we do not always create a separate AppDomain for doing the ReflectionOnlyLoad.

To verify the need,  I checked all the places where we use it and looked at the impact. Here's what I found:

TestEngineActivator (nunit.engin.api)
Loads all engines installed and checks the version. At the point where we release NUnit 3.0, there will only be one engine version to load, so it is not an immediate problem. The engine loaded in the RO context will also be loaded in an execution context. However, once we release version 3.2 of the engine, it is certainly possible that an extra engine will remain in memory and as further releases ensue, we could end up with a bunch of them. 

DriverService (nunit.engine and nunit.core.engine)
The service loads any potential test assembly into a RO context. There are two cases. Either a driver is found for it or not. If a driver is found, then it will be immediately loaded into an execution context. If no driver is found, it's an execution error. When running under the console, the process terminates immediately, so there is no memory issue. When running under the test adapter it's a slight problem because VS does call NUnit with non-test assemblies. For a future gui, it would also present a problem.

Mono.Addins does not do a ReflectionOnlyLoad. There are three calls in a block that is commented out. My guess is that they had a memory leak. They switched it to LoadFile, which is an odd sort of call. Here's a blog post I found about it: http://blogs.msdn.com/b/suzcook/archive/2003/09/19/loadfile-vs-loadfrom.aspx

For Mono.Addins, it's a bit academic since it's not our code. Additionally, there's a chance we will stop using it in the future - we don't get a lot from it considering how much trouble it is. More on this in the next cycle.

In addition to the ReflectionOnlyLoads we use, we also use a file-based approach to looking at Assemblies. The class AssemblyReader is used to read an assembly into memory as a file and examine the various headers. I think I took the code from a C program I wrote a long time ago. I had plans to expand this eventually to replace some of the ReflectionOnlyLoads. We use AssemblyReader in the following places:

RuntimeFrameworkService (nunit.engine and nunit.core.engine)
The service looks at what framework target was used in building the image and whether it is 32 or 64 bits so we can start a process using the correct version of .NET and the correct platform. FYI, this service will probably be rewritten to handle a single assembly rather than a package with multiple assemblies. The decision about how to handle multiple assemblies with different runtime versions is a higher-level policy decision and doesn't belong in here.

So, I generally like the idea, but I'd like you should look at what we already have and considering cleaning it up and extending it to get the various info bits we need. It adds far less than 30K to our assembly and I assume it already works in CF.

[Just read further and I see you already found AssemblyReader.cs.]

As far as priority goes, here are my thoughts, per use case and in summary:

 * TestEngineActivator - Low priority right now, high for the 3.2 release.
 * DriverService - Normal priority right now, high when we have a GUI.
 * RuntimeFrameworkService - Low priority until we do it for some other reason.

In sum, I think it's normal priority right now and at some point will be high. If it's something you want to spend time on right now, then I would bump it to normal and try to get it in the next milestone.

As usual, please self-assign it if you want to work on it now, but leave it unassigned if you think you won't work on it right away.

**Comment by oznetmaster** [comment]
I think you will find that the code required to get the assembly references is non-trivial.  That is why I went to an open source library to do the work (mono.cecil and then asmex.fileviewer).  They both contain all the code required.  I have been sliming the one that I am currently using (asmex.fileviewer) down, so it will be smaller.  I also added Lazy initialization to it so minimal work is being done except where actually used.

 

We could roll our own, but the development and debugging time may not be worth the effort.

**Comment by CharliePoole** [comment]
Fair enough. We'll need to consider licensing issues.

Mono.cecil has the advantage that it's MIT/X11, is developed continuously by guys I know and trust it has a nuget package. It could also be used to replace some other reflection-based stuff if we used it in it's entirety. For the desktop, we could just pull in the package. I can guess, however, that you may have had to modify some things to make it work in CF. How big is it?

I'm not sure if I have found the right code for asmex.fileviewer. Is it this? 
   http://www.codeproject.com/Articles/3262/A-NET-assembly-viewer
If that's the one, it doesn't look too attractive. There is no explicit license so we take our chances. Granted, it's probably safe and we could always ask the author to contribute it. The year of the article is 2002 and I don't find it being maintained anywhere. Of course, I may not be looking at the right code.

**Comment by oznetmaster** [comment]
Yes, that is the one.

 

It seemed from reading the copyright that it was a pretty open “license”.  It would seem that approaching the author to contribute it would make sense.  Only six of the source files are actually required.  It has not been maintained, but works perfectly, both in CF and normal .NET

 

As to Mono.Cecil, I have had the full library (READ_ONLY) working under CF for a long time.  That is a pretty large assembly (210KB).  The one I am using in the CF core engine is stripped down from that (I call it miniCecil).   Still a lot larger than the asmex library.  It may be I could strip more from cecil.  Have not spent a lot of time trying to optimize it.

**Comment by oznetmaster** [comment]
The authors own website (jbrowse.com) is defunt.  His company’s (riskcare.com) appears to still be there.  It is a pretty large company, with offices in London, New York and Sydney.  You might be the better person to contact them and see if they would contribute it.  They probably no longer have a real vested interest in it after all this time. http://www.riskcare.com

**Comment by oznetmaster** [comment]
What makes cecil so large is that it attempts to mimic Syste.Reflection and System.Reflection.Emit in total.  Even the READ_ONLY build attempts to mimic all of System.Reflection.

 

All NUnit needs is information extraction.  The asmex library is only an information retrieval library, so it starts a lot smaller and simpler.  By adding Lazy initialization, I have minimized startup overhead and total memory utilization.  It would seem the better choice if the licensing could be clarified.

**Comment by CharliePoole** [comment]
@oznetmaster May I suggest that if you are going to reply by email to a github issue or pr, you remove the copies of earlier mails and posts. It makes the issue form virtually unreadable to have each comment get longer and longer and contain a copy of all previous comments. Alternatively, click the link and reply on GitHub itself.

**Comment by CharliePoole** [comment]
For mono.cecil: How big is the Readonly build?

For asmex.fileviewer: The copyright holder is a large consultancy company. I'd rather not get involved with them and - as it would probably turn out - their legal department. So if we use this it will have to be under the terms of the license included in the code. The relevant statements are:

 * You are hereby granted the right to redistribute this source unmodified
 * in its original archive. 

We wouldn't be doing that, so it's not applicable.

 * You are hereby granted the right to use this code, or code based on it,
 * provided that you acknowledge RiskCare Ltd somewhere in the documentation
 * of your application. 

We would be doing that. We'll need to include them in the NOTICES.txt file, which is not a problem.

 * You are hereby granted the right to distribute changes to this source, 
 * provided that:
 * 
 * 1 -- This copyright notice is retained unchanged 
We would be doing that. We would have to keep the copyright notice on each file and not replace it with our own.

 * 2 -- Your changes are clearly marked 
This is pretty burdensome. Every line or block of code we ever change in any of these files will have to have a comment to indicate that you changed it.

Charlie

**Comment by oznetmaster** [comment]
The CF mono.cecil READ_ONLY build is 210KB.

 

The only reason I thought it might be easier to get them to contribute is the age of the code, and the fact that it was originally written as a program for use in their training department.  It was never part of their product line:

 

Asmex was an educational project; the idea was to make an application that involved knowledge of the very lowest possible level of .NET, yet also took advantage of the clean GUI model of WinForms. It is used in  <http://www.riskcare.com/> our company for training and debugging purposes.

**Comment by oznetmaster** [comment]
Perhaps we could ask them to simply change the license to be the MIT license?  That might be easier?  I have never dealt with this kind of thing before.

**Comment by CharliePoole** [comment]
I notice that svg.net, published by the same author in an article the following year, has the two-clause BSD license. Rafael Teixeira has re-booted the project as a part of his managed commons: http://github.com/monoman/svgnet. In his source, you can see how he re-licensed it under MIT/X11 but included a notice that conforms to the BSD terms. Perhaps we could do something similar. I'll sleep on it.

**Comment by oznetmaster** [comment]
This may be a non-issue.

 

I have my mini cecil down to about 50KB on the CF build, and have more possible optimizations.  So using cecil may be the best option.

**Comment by CharliePoole** [comment]
Sounds like that may be the way to go, especially as it uses the same license we do.

**Comment by oznetmaster** [comment]
The mini-cecil that I am using has very limited capabilities.  Information gathering only, and then only about assemblies and modules.  If you want to use it for more reflection kinds of things, it will get larger.

**Comment by CharliePoole** [comment]
Not a problem. We can live with what is there and put back what we want when we want it. 

It should start out in our repo identical to the mono version and then get modified as needed so that it remains mergeable with any fixes they do.

Were you going to take on this issue?

**Comment by oznetmaster** [comment]
Since I already have done most of the work, it makes sense.

 

The unmodified mono version of cecil already compiles for a CF version, (CF build), so it is easy to start there and remove/add what we need.  Just starting with a READ_ONLY, CF, NO_CODE, NO_MEMBERS build is about 80% of the way there, with no modifications.

 

Should we put it in its own folder in the repo? Or somewhere else?  I assume we will build it as its own assembly?

**Comment by CharliePoole** [comment]
@oznetmaster I was asking because you haven't assigned yourself. Never mind, I just did it. I'll leave it up to you to give it a milestone.

Since I'll be out of things for a month, I'll leave it up to you and @rprouse to figure out where to put it. If we think it will be used by other people for other things besides NUnit, we could make it a separte repo. Otherwise, keep it as a part of nunit.

You'll want to add a new project to the solution. The easiest way is to create a new directory for it under NUnitEngine and put it there. Then just add an existing project to the solution. As I say, I'd start with the original from mono and then make changes in git so there is a history.

**Comment by oznetmaster** [comment]
There seems to be good arguments for both approaches.

 

If we make our version of cecil its own repo, then we can start by forking the existing cecil repo.  This would make it easier to keep our version in sync with the normal version.  This would, however, seemingly complicate the build and release process?  Not sure how we would handle that.

 

If we include cecil as a project in the NUnit solution (both regular and CF), then build/distribution is trivial, but keeping our version in sync with the normal version seemingly becomes more difficult?

 

Any input?

**Comment by CharliePoole** [comment]
If we make it it's own repo and build it, then we can create a nuget package and include it in most of our projects. That won't work so well for VS2008, however. A submodule is another possibility, but it would be our first use of this technique so is a little risky.

@rprouse I'll leave this for you to figure out with Neil.

**Comment by oznetmaster** [comment]
I am not sure what a submodule is?

**Comment by CharliePoole** [comment]
It's a git thing allowing source to be shared between repositories. Some people swear by it. Other people warn against it. I've never used it and I don't know if @rprouse has. We may want to create a sort of sandbox repo to experiment if we decide to try this.

**Comment by CharliePoole** [comment]
I have changed this to milestone beta-4 because I think we need to resolve it sooner rather than later.

@oznetmaster I guess we had a misunderstanding about your taking on the issue - I removed your name and added mine as the assignee. Please keep contributing your ideas.

I think my next step is to create a spike that uses cecil.

**Comment by oznetmaster** [comment]
I never said I would not take this on.  I said that I did not have the git expertise required to either fork or to create a sub-repo, or whatever is the best structure to maintain our version of cecil.  I had asked that some else undertake that specific step.  Once that is done, I will port all of my conditionals to the code so different minimal versions of it can be created

**Comment by CharliePoole** [comment]
As I said, it was my misunderstanding. Rob created a repo many months back,
before my vacation trip. Nothing has been done in it except to pull in
changes from the jbevain/cecil repo.

I'll use this to do a spike that will hopefully enable us to make a
decision about using it before we try to minimize the size.

Charlie

**Comment by oznetmaster** [comment]
Where is that repo?

Aside from minimizing the size, the CF conditionals need to be updated, as well as the readonly ones.

My CF cecil is also fully up to date with all of the jbevain/cecil repo, but with all of the CF, readonly, and other size minimization conditionals in it.

**Comment by CharliePoole** [comment]
It's under the nunit organization, I.E. a sibling of nunit/nunit.

**Comment by oznetmaster** [comment]
I am probably just ignorant, but it seems like i cannot clone both the main branch and the nunit fork in my machine at the same time.  They both appear in my list of repos in github, but  they keep showing up as the same identical set of files.

Is there a secret to doing this.

**Comment by rprouse** [comment]
@oznetmaster, sorry, but I am not sure what you are saying there. If you are having trouble cloning our copy of Cecil, go to https://github.com/nunit/cecil, it is a different repository.

If you are trying to clone nunit/nunit and your fork, oznetmaster/nunit, it should be fairly straightforward at the command line.

What are you using to clone repositories and exactly which repositories are you having trouble with?

**Comment by CharliePoole** [comment]
If you want to work with both versions on your machine, then you have to use a separate working directory. There's no need to go through github to do it, just make a local clone, just as we usually do when working with NUnit.

**Comment by oznetmaster** [comment]
The separate working directories appears to be the issue.  I have all of my github clone repos in a single folder.  I will need to do something different for the nunit/cecil one.

**Comment by CharliePoole** [comment]
But each repo is a subfolder, so simply give it a different name.

What command do you enter to clone a repo?

**Comment by rprouse** [comment]
The easiest thing to do is rename your clone if Cecil in your directory to something like CecilOz then clone the nunit copy. 

The correct way would be to add nunit as a new remote to your copy and map nunit master to a branch on the clone. This would allow you to merge across branches, but requires more git knowledge (or googling). If you want to do that, I can give you the commands, but you will need to be careful which remote and which branch you push to.

**Comment by CharliePoole** [comment]
I'm closing this issue, since we have implemented use of Cecil for all versions of the engine we now create, including the core engine used by the adapter. If we decide to create other engines (for CF, for example) then we can deal with that in a separate issue.
